src/openstack_manager.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for handling interactions with OpenStack."""
+import logging
+from pathlib import Path
+
+import keystoneauth1.exceptions
+import openstack
+import openstack.exceptions
+import yaml
+from openstack.identity.v3.project import Project
+
+from errors import OpenStackInvalidConfigError, OpenStackUnauthorizedError
+
+logger = logging.getLogger(__name__)
+
+CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
+
+
+def _validate_cloud_config(cloud_config: dict) -> None:
+    """Validate the format of the cloud configuration.
+
+    Args:
+        cloud_config: The configuration in clouds.yaml format to validate.
+
+    Raises:
+        InvalidConfigError: if the format of the config is invalid.
+    """
+    # dict of format: {clouds: <cloud-name>: <cloud-config>}
+    try:
+        clouds = list(cloud_config["clouds"].keys())
+    except KeyError as exc:
+        raise OpenStackInvalidConfigError("Invalid clouds.yaml.") from exc
+
+    if not clouds:
+        raise OpenStackInvalidConfigError("No clouds defined in clouds.yaml.")
+
+
+def _write_config_to_disk(cloud_config: dict) -> None:
+    """Write the cloud configuration to disk.
+
+    Args:
+        cloud_config: The configuration in clouds.yaml format to write to disk.
+    """
+    CLOUDS_YAML_PATH.parent.mkdir(parents=True, exist_ok=True)
+    with CLOUDS_YAML_PATH.open("w", encoding="utf-8") as clouds_yaml:
+        yaml.dump(cloud_config, clouds_yaml)
+
+
+def _create_connection(cloud_config) -> openstack.connection.Connection:
+    """Create a connection object.
+
+    Args:
+        cloud_config: The configuration in clouds.yaml format to apply.
+    Raises:
+        InvalidConfigError: if the config has not all required information.
+
+    Returns:
+        An openstack.connection.Connection object.
+    """
+    clouds = list(cloud_config["clouds"].keys())
+    if len(clouds) > 1:
+        logger.warning("Multiple clouds defined in clouds.yaml. Using the first one to connect.")
+    cloud_name = clouds[0]
+
+    # api documents that keystoneauth1.exceptions.MissingRequiredOptions can be raised but
+    # I could not reproduce it. Therefore, no catch here.
```

Comment:
Perhaps worth checking with the team? If the documentation says it can be raised we should probably handle it

Reply:
Just tried reproducing with invalid `clouds.yaml` (commented out fields, deleted fields) - they don't raise this exception. 
```
clouds:
  sunbeam:
    auth:
      auth_url: http://10.20.21.12/openstack-keystone
#      password: 8gfWnW157KJ7
#      project_domain_name: admin_domain
#      project_name: admin
#      user_domain_name: admin_domain
#      username: admin
```
Raised an [issue on the bugtracker](https://bugs.launchpad.net/openstacksdk/+bug/2054361).